src/examples/4626-router/Generic4626Router.sol

```diff
@@ -0,0 +1,229 @@
+            // We enforce pool initialization via the hook, this way we can
+            // ensure that the pool is initialized with the correct parameters
+            revert Generic4626Router__NotAllowed();
+        }
+
+        return IHooks.beforeInitialize.selector;
+    }
+
+    function _beforeAddLiquidity(
+        address,
+        PoolKey calldata,
+        IPoolManager.ModifyLiquidityParams calldata,
+        bytes calldata
+    ) internal pure override returns (bytes4) {
+        revert Generic4626Router__NotAllowed();
+    }
+
+    function _beforeSwap(
+        address,
+        PoolKey calldata poolKey,
+        IPoolManager.SwapParams calldata params,
+        bytes calldata
+    )
+        internal
+        override
+        returns (bytes4 selector, BeforeSwapDelta swapDelta, uint24 lpFeeOverride)
+    {
+        PoolId poolId = poolKey.toId();
+        PoolDetails memory details = poolDetails[poolId];
+
+        bool isExactInput = params.amountSpecified < 0;
+
+        // This following section is hard to follow, I initially wrote this with currency0 and currency1
+        // but it was truly impossible to follow. Ended up storing the vault and underlying.
+        // Although, I think I got a hang of it so might go back
+        if (params.zeroForOne == details.wrapDirection) {
+            uint256 inputAmount = isExactInput
+                ? uint256(-params.amountSpecified)
+                : _getUnderlyingForShares(details.vault, uint256(params.amountSpecified));
+
+            CurrencySettler.take(
+                Currency.wrap(address(details.underlying)),
+                poolManager,
+                address(this),
+                inputAmount,
+                false
+            );
+
+            uint256 shares = _deposit(details.underlying, details.vault, inputAmount);
+
+            CurrencySettler.settle(
+                Currency.wrap(address(details.vault)), poolManager, address(this), shares, false
+            );
+
+            int128 amountUnspecified =
+                isExactInput ? -shares.toInt256().toInt128() : inputAmount.toInt256().toInt128();
+
+            swapDelta = toBeforeSwapDelta(-params.amountSpecified.toInt128(), amountUnspecified);
+        } else {
+            uint256 inputAmount = isExactInput
+                ? uint256(-params.amountSpecified)
+                : _getSharesForUnderlying(details.vault, uint256(params.amountSpecified));
+
+            CurrencySettler.take(
+                Currency.wrap(address(details.vault)),
+                poolManager,
+                address(this),
+                inputAmount,
+                false
+            );
+
+            uint256 underlyingAmount = _withdraw(details.vault, inputAmount);
+
+            CurrencySettler.settle(
+                Currency.wrap(address(details.underlying)),
+                poolManager,
+                address(this),
+                underlyingAmount,
+                false
+            );
```

Comment:
optimizable -- shares should be burned and underlying should be sent directly to PoolManager

Reply:
```solidity
poolManager.sync(vault.asset);
vault.redeem(...); // redeem underlying ERC20 to poolManager
poolManager.settle();
```

Reply:
This one is interesting, does seem like a worthwhile optimization.